092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,253 @@
+
+#### Notifiers in Cruise Control
+
+Whenever anomalies are detected, Cruise Control provides the ability to notify the user regarding the detected anomalies using optional notifier classes.
+The notification sent by these classes increases the visibility of the operations that are taken by Cruise Control.
+However, the notifier class used by Cruise Control is configurable and custom notifiers can be used by setting the `anomaly.notifier.class` property.
+The notifier class returns the `action` that is going to be taken on the flagged anomaly.
+These actions can be classified into three types:
+* FIX - Start the anomaly fix
+* CHECK - Delay the anomaly fix
+* IGNORE - Ignore the anomaly fix
+
+The default `NoopNotifer` always sets the notifier action as `IGNORE` means that the detected anomaly will be silently ignored.
+
+Cruise Control also provides [custom notifiers](https://github.com/linkedin/cruise-control/wiki/Configure-notifications) like Slack Notifier, Alerta Notifier etc. for notifying users regarding the anomalies. There are multiple other [self-healing notifier](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations) related configurations you can use to make notifiers more efficient as per the use case.
+
+#### Self Healing
+
+If self-healing is enabled in Cruise Control and anomalies are detected, then a fix for these will be attempted automatically.
+The action returned by the notifier would decide whether the anomaly should be fixed or not.
+If the notifier has returned `FIX` as the action then the classes which are responsible for resolving the anomaly would be called.
+Each detected anomaly is implemented by a specific class which leverages a corresponding other class to run a fix.
+For example, the `GoalViolations` class uses the `RebalanceRunnable` class, the `DiskFailure` class use the `RemoveDisksRunnable` class and so on.
+An optimization proposal will then be generated by these `Runnable` classes and that proposal will be applied on the cluster to fix the anomaly.
+In case the anomaly detected is unfixable for e.g. violated hard goals that cannot be fixed typically due to lack of physical hardware(insufficient number of racks to satisfy rack awareness, insufficient number of brokers to satisfy Replica Capacity Goal, or insufficient number of resources to satisfy resource capacity goals), the anomaly wouldn't be fixed and the cruise control logs will be updated with `self-healing is not possible due to unfixable goals` warning.
+When self-healing starts, you can check the anomaly status by polling the [`state` endpoint](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) with `substate` parameter set as `anomaly_detector`. This would provide the anomaly details like anomalyId, the type of anomaly and the status of the anomaly etc.
```

Comment:
```suggestion
When self-healing starts, you can check the anomaly status by querying the [`state` endpoint](https://github.com/linkedin/cruise-control/wiki/REST-APIs#query-the-state-of-cruise-control) with `substate` parameter set as `anomaly_detector`. This would provide the anomaly details like anomalyId, the type of anomaly and the status of the anomaly etc.
```